io/pci/pci_hotplug.py

```diff
@@ -47,76 +47,88 @@ def setUp(self):
-                                      % self.device)
-            self.slot = genio.read_file("/proc/device-tree/%s/ibm,loc-code"
-                                        % devspec)
-            self.slot = re.match(r'((\w+)[\.])+(\w+)-P(\d+)-C(\d+)|Slot(\d+)',
-                                 self.slot).group()
+                                      % pci_addr)
+            slot = genio.read_file("/proc/device-tree/%s/ibm,loc-code"
+                                   % devspec)
+            slot = re.match(r'((\w+)[\.])+(\w+)-P(\d+)-C(\d+)|Slot(\d+)',
+                            slot).group()
         else:
-            self.slot = pci.get_pci_prop(self.device, "PhySlot")
-        if not os.path.isdir('/sys/bus/pci/slots/%s' % self.slot):
-            self.cancel("%s Slot not available" % self.slot)
-        if not os.path.exists('/sys/bus/pci/slots/%s/power' % self.slot):
-            self.cancel("%s Slot does not support hotplug" % self.slot)
+            slot = pci.get_pci_prop(pci_addr, "PhySlot")
+        if not os.path.isdir('/sys/bus/pci/slots/%s' % slot):
+            self.cancel("%s Slot not available" % slot)
+        if not os.path.exists('/sys/bus/pci/slots/%s/power' % slot):
+            self.cancel("%s Slot does not support hotplug" % slot)
+        return slot
 
     def test(self):
         """
         Creates namespace on the device.
         """
+        err_pci = []
+        for pci_addr in self.device.split(","):
+            if not os.path.isdir('/sys/bus/pci/devices/%s' % pci_addr):
+                self.err_pci.append(pci_addr)
+                continue
+            slot = self.get_slot(pci_addr)
+            for _ in range(self.count):
+                if not self.hotplug_remove(slot, pci_addr):
+                    err_pci.append(pci_addr)
+                else:
+                    self.log.info("%s removed successfully" % pci_addr)
+                if not self.hotplug_add(slot, pci_addr):
+                    err_pci.append(pci_addr)
+                else:
+                    self.log.info("%s added back successfully" % pci_addr)
+        if err_pci:
+            self.fail("following devices not found: %s" % err_pci)
 
-        for _ in range(self.num_of_hotplug):
-            self.hotplug_remove()
-            self.hotplug_add()
-            self.check_add_remove()
-
-    def hotplug_remove(self):
+    def hotplug_remove(self, slot, pci_addr):
         """
         Hot Plug remove operation
         """
-        genio.write_file("/sys/bus/pci/slots/%s/power" % self.slot, "0")
-        time.sleep(5)
-        if self.device in pci.get_pci_addresses():
-            self.return_code = 1
-        else:
-            self.log.info("Adapter %s removed successfully", self.device)
+        genio.write_file("/sys/bus/pci/slots/%s/power" % slot, "0")
+
+        def is_removed():
+            if pci_addr in pci.get_pci_addresses():
+                return False
+            return True
```

Comment:
@Naresh-ibm Why do you want this to be a function inside another function ?

Reply:
wait_for function has this syntax and I felt it is easy to undestand

Reply:
Does wait_for mandate to call a function only within its scope ?

Reply:
what is the other way your expecting?

Reply:
Is this possible:

```
    def is_removed():
        if pci_addr in pci.get_pci_addresses():
            return False
        return True

    def hotplug_remove(self, slot, pci_addr):
        """
        Hot Plug remove operation
        """
        genio.write_file("/sys/bus/pci/slots/%s/power" % slot, "0")
        return wait.wait_for(is_removed, timeout=10) or False
```

Reply:
having a nested function is always better than a separate function @narasimhan-v

Reply:
@Naresh-ibm This is new to me, any more explanations or pointers for this please ?

Reply:
check multipath library which was accepted there

Reply:
@Naresh-ibm 
1. I asked if we can have that checking function outside. Yes / No ?
2. The reason I asked is because: You can have same function for multiple checks there. Code duplication is reduced and so better maintenance.
3. I have no reason to assume nested functions are better, just because it is used in multipath utility. And no reason you shouldn't use it either. All I wanted is a proof of why it was better, since you had said so. It can be better, but I just wanted to know why/how it is better :)

Reply:
> @Naresh-ibm
> 
>     1. I asked if we can have that checking function outside. Yes / No ?
> 
=====>>> why not obviously you can have it outside, but return types not usefull  for all condition check and hence ultimately we have add both the nested function together.

>     2. The reason I asked is because: You can have same function for multiple checks there. Code duplication is reduced and so better maintenance.
> 

====>>> we already had the separate function previously and you said to get the result from skipping the existing check function in our last FTF discussion. 
when I was doing the same I found the sleep time between the operations, so I removed the sleep time and added the nested function which I feel the best way

I understand your concern of code duplicate but if you write the the separate function also the same steps from both the nested function will have to combined to achieve the same.
>     3. I have no reason to assume nested functions are better, just because it is used in multipath utility. And no reason you shouldn't use it either. All I wanted is a proof of why it was better, since you had said so. It can be better, but I just wanted to know why/how it is better :)

this is mainly to overcome the waiting time that we used to face using sleep time. 
==>> this is better  in following way(as per my knowledge)
1. scopes will be different 
  a. separate function will get class level global scope but nested will complete the task with local scope which good according to coding standards
2. work flow wise and ease of use is good with nested function.
3. any new user can easily get the info of next steps if all operations are at one place  instead of navigating between the functions (possible with nested functions)

as coder and user I feel these are better. I am  just putting my view here.
let me know if you still feel separate function is better than what I did and how can we achieve it

Reply:
> I understand your concern of code duplicate but if you write the the separate function also the same steps from both the nested function will have to combined to achieve the same.

For this reason, I am fine with this approach.

> separate function will get class level global scope but nested will complete the task with local scope which good according to coding standards

There is nothing in documentation that I can find which supports this. So let us keep this discussion aside.